attribute/hash_test.go

```diff
@@ -0,0 +1,161 @@
+	func(k string) KeyValue { return String(k, "foo") },
+	func(k string) KeyValue { return String(k, "bar") },
+	func(k string) KeyValue { return StringSlice(k, []string{"foo", "bar", "baz"}) },
+	func(k string) KeyValue { return StringSlice(k, []string{"[]i1"}) },
+}
+
+func TestHashKVsEquality(t *testing.T) {
+	type testcase struct {
+		hash fnv.Hash
+		kvs  []KeyValue
+	}
+
+	keys := []string{"k0", "k1"}
+
+	// Test all combinations up to length 3.
+	n := len(keyVals)
+	result := make([]testcase, 0, 1+len(keys)*(n+(n*n)+(n*n*n)))
+
+	result = append(result, testcase{hashKVs(nil), nil})
+
+	for _, key := range keys {
+		for i := 0; i < len(keyVals); i++ {
+			kvs := []KeyValue{keyVals[i](key)}
+			hash := hashKVs(kvs)
+			result = append(result, testcase{hash, kvs})
+
+			for j := 0; j < len(keyVals); j++ {
+				kvs := []KeyValue{
+					keyVals[i](key),
+					keyVals[j](key),
+				}
+				hash := hashKVs(kvs)
+				result = append(result, testcase{hash, kvs})
+
+				for k := 0; k < len(keyVals); k++ {
+					kvs := []KeyValue{
+						keyVals[i](key),
+						keyVals[j](key),
+						keyVals[k](key),
+					}
+					hash := hashKVs(kvs)
+					result = append(result, testcase{hash, kvs})
+				}
+			}
+		}
+	}
+
+	for i := 0; i < len(result); i++ {
+		hI, kvI := result[i].hash, result[i].kvs
+		for j := 0; j < len(result); j++ {
+			hJ, kvJ := result[j].hash, result[j].kvs
+			m := msg{i: i, j: j, hI: hI, hJ: hJ, kvI: kvI, kvJ: kvJ}
+			if i == j {
+				m.cmp = "=="
+				if hI != hJ {
+					t.Errorf("hashes not equal: %s", m)
+				}
+			} else {
+				m.cmp = "!="
+				if hI == hJ {
+					// Do not use testify/assert here. It is slow.
+					t.Errorf("hashes equal: %s", m)
+				}
+			}
+		}
+	}
+}
+
+type msg struct {
+	cmp      string
+	i, j     int
+	hI, hJ   fnv.Hash
+	kvI, kvJ []KeyValue
+}
+
+func (m msg) String() string {
+	return fmt.Sprintf(
+		"(%d: %d)%s %s (%d: %d)%s",
+		m.i, m.hI, slice(m.kvI), m.cmp, m.j, m.hJ, slice(m.kvJ),
+	)
+}
+
+func slice(kvs []KeyValue) string {
+	if len(kvs) == 0 {
+		return "[]"
+	}
+
+	var b strings.Builder
+	_, _ = b.WriteRune('[')
+	_, _ = b.WriteString(string(kvs[0].Key))
+	_, _ = b.WriteRune(':')
+	_, _ = b.WriteString(kvs[0].Value.Emit())
+	for _, kv := range kvs[1:] {
+		_, _ = b.WriteRune(',')
+		_, _ = b.WriteString(string(kv.Key))
+		_, _ = b.WriteRune(':')
+		_, _ = b.WriteString(kv.Value.Emit())
+	}
+	_, _ = b.WriteRune(']')
+	return b.String()
+}
+
+var h fnv.Hash
+
+func BenchmarkHashKVs(b *testing.B) {
+	attrs := make([]KeyValue, len(keyVals))
+	for i := range keyVals {
+		attrs[i] = keyVals[i]("k")
+	}
+
+	b.ResetTimer()
+	b.ReportAllocs()
+	for range b.N {
+		h = hashKVs(attrs)
+	}
+}
+
+func FuzzHashKVs(f *testing.F) {
+	f.Fuzz(func(_ *testing.T, k1, k2, k3, k4, k5, s string, i int, i64 int64, f float64, b bool) {
+		kvs := []KeyValue{
+			String(k1, s),
+			Int(k2, i),
+			Int64(k3, i64),
+			Float64(k4, f),
+			Bool(k5, b),
+		}
+		h = hashKVs(kvs)
```

Comment:
this is not a benchmark

```suggestion
		hashKVs(kvs)
```